File: zone_filler.rb


class ZoneFiller

  attr_reader :grid

  # if a grid is given, it is used as starting point; 
  # otherwise, the goban scoring_grid is used.
  def initialize(goban, grid=nil)
    grid = goban.scoring_grid.convert(goban.grid) if !grid
    # $log.debug("ZoneFiller.new \n"+grid.to_s) if $debug
    @goban = goban
    @grid = grid
    @yx = grid.yx
    @groups = nil
  end

  # "Color" a goban zone.
  # to_replace can be EMPTY or a zone code (but cannot be a real color like BLACK)
  # neighbors, if given should be an array of n arrays, with n == number of colors
  # if neighbors are not given, we do simple "coloring"
  def fill_with_color(start_i, start_j, to_replace, color, neighbors=nil)
    # $log.debug("fill #{start_i} #{start_j}; replace #{to_replace} with #{color}") if $debug
    return 0 if @yx[start_j][start_i] != to_replace
    size = 0
    @to_replace = to_replace
    @groups = neighbors
    gaps = [[start_i, start_j, start_j]]
    while (gap = gaps.pop)
      # $log.debug("About to do gap: #{gap} (left #{gaps.size})") if $debug
      i,j0,j1 = gap
      next if @yx[j0][i] != to_replace # gap already done by another path
      while _check(i,j0-1) do j0 -= 1 end
      while _check(i,j1+1) do j1 += 1 end
      size += j1-j0+1
      # $log.debug("Doing column #{i} from #{j0}-#{j1}") if $debug
      (i-1).step(i+1,2) do |ix|
        curgap = nil
        j0.upto(j1) do |j|
          # $log.debug("=>coloring #{i},#{j}") if $debug and ix<i
          @yx[j][i] = color if ix<i
          # $log.debug("checking neighbor #{ix},#{j}") if $debug
          if _check(ix,j)
            if ! curgap
              # $log.debug("New gap in #{ix} starts at #{j}") if $debug
              curgap = j # gap start
            end
          else
            if curgap
              # $log.debug("--- pushing gap [#{ix},#{curgap},#{j-1}]") if $debug
              gaps.push([ix,curgap,j-1])
              curgap = nil
            end
          end
        end # upto j
        # $log.debug("--- pushing gap [#{ix},#{curgap},#{j1}]") if $debug and curgap
        gaps.push([ix,curgap,j1]) if curgap # last gap
      end # each ix
    end # while gap
    return size
  end

private

  # Returns true if the replacement is needed (=> i,j has a color equal to the replaced one)
  def _check(i,j)
    color = @yx[j][i]
    return false if color == BORDER
    return true if color == @to_replace
    if @groups and color < 2
      group = @goban.stone_at?(i,j).group
      @groups[color].push(group) if group and ! @groups[color].find_index(group)
    end
    return false
  end

end

File: js/ZoneFiller.js
//Translated from zone_filler.rb using babyruby2js
'use strict';

var main = require('./main');

/** @class public read-only attribute: grid
 *  if a grid is given, it is used as starting point; 
 *  otherwise, the goban scoring_grid is used.
 */
function ZoneFiller(goban, grid) {
    if (grid === undefined) grid = null;
    if (!grid) {
        grid = goban.scoringGrid.convert(goban.grid);
    }
    this.goban = goban;
    this.grid = grid;
    this.yx = grid.yx;
    this.groups = null;
}
module.exports = ZoneFiller;

// "Color" a goban zone.
// to_replace can be EMPTY or a zone code (but cannot be a real color like BLACK)
// neighbors, if given should be an array of n arrays, with n == number of colors
// if neighbors are not given, we do simple "coloring"
ZoneFiller.prototype.fillWithColor = function (startI, startJ, toReplace, color, neighbors) {
    if (neighbors === undefined) neighbors = null;
    // $log.debug("fill #{start_i} #{start_j}; replace #{to_replace} with #{color}") if $debug
    if (this.yx[startJ][startI] !== toReplace) {
        return 0;
    }
    var vcount = 0;
    this.toReplace = toReplace;
    this.groups = neighbors;
    var gaps = [[startI, startJ, startJ]];
    var gap, i, j0, j1;
    while ((gap = gaps.pop())) {
        // $log.debug("About to do gap: #{gap} (left #{gaps.size})") if $debug
        var _m = gap;
        i = _m[0];
        j0 = _m[1];
        j1 = _m[2];
        
        if (this.yx[j0][i] !== toReplace) { // gap already done by another path
            continue;
        }
        while (this._check(i, j0 - 1)) {
            j0 -= 1;
        }
        while (this._check(i, j1 + 1)) {
            j1 += 1;
        }
        vcount += j1 - j0 + 1;
        // $log.debug("Doing column #{i} from #{j0}-#{j1}") if $debug
        for (var ix = (i - 1); ix <= i + 1; ix += 2) {
            var curgap = null;
            for (var j = j0; j <= j1; j++) {
                // $log.debug("=>coloring #{i},#{j}") if $debug and ix<i
                if (ix < i) {
                    this.yx[j][i] = color;
                }
                // $log.debug("checking neighbor #{ix},#{j}") if $debug
                if (this._check(ix, j)) {
                    if (!curgap) {
                        // $log.debug("New gap in #{ix} starts at #{j}") if $debug
                        curgap = j; // gap start
                    }
                } else if (curgap) {
                    // $log.debug("--- pushing gap [#{ix},#{curgap},#{j-1}]") if $debug
                    gaps.push([ix, curgap, j - 1]);
                    curgap = null;
                }
            }
            // upto j
            // $log.debug("--- pushing gap [#{ix},#{curgap},#{j1}]") if $debug and curgap
            if (curgap) { // last gap
                gaps.push([ix, curgap, j1]);
            }
        } // each ix
    }
    // while gap
    return vcount;
};

//private;
// Returns true if the replacement is needed (=> i,j has a color equal to the replaced one)
ZoneFiller.prototype._check = function (i, j) {
    var color = this.yx[j][i];
    if (color === main.BORDER) {
        return false;
    }
    if (color === this.toReplace) {
        return true;
    }
    if (this.groups && color < 2) {
        var group = this.goban.stoneAt(i, j).group;
        if (group && this.groups[color].indexOf(group) < 0) {
            this.groups[color].push(group);
        }
    }
    return false;
};

// E02: unknown method: find_index(...)